src/include/daos/rpc.h

```diff
@@ -40,23 +40,46 @@
 	 (rpc_ver & RPC_VERSION_MASK) << RPC_VERSION_OFFSET |	\
 	 (mod_id & MODID_MASK) << MODID_OFFSET)
 
+#define DAOS_MODULE_LIST                                                                           \
+	X(DAOS_VOS_MODULE, 0)     /** version object store */                                      \
+	X(DAOS_MGMT_MODULE, 1)    /** storage management */                                        \
+	X(DAOS_POOL_MODULE, 2)    /** pool service */                                              \
+	X(DAOS_CONT_MODULE, 3)    /** container service */                                         \
+	X(DAOS_OBJ_MODULE, 4)     /** object service */                                            \
+	X(DAOS_REBUILD_MODULE, 5) /** rebuild **/                                                  \
+	X(DAOS_RSVC_MODULE, 6)    /** replicated service server */                                 \
+	X(DAOS_RDB_MODULE, 7)     /** rdb */                                                       \
+	X(DAOS_RDBT_MODULE, 8)    /** rdb test */                                                  \
+	X(DAOS_SEC_MODULE, 9)     /** security framework */                                        \
+	X(DAOS_DTX_MODULE, 10)    /** DTX */                                                       \
+	X(DAOS_PIPELINE_MODULE, 11)                                                                \
+	X(DAOS_NR_MODULE, 12)  /** number of defined modules */                                    \
+	X(DAOS_MAX_MODULE, 64) /** Size of uint64_t see dmg profile */
+
 enum daos_module_id {
-	DAOS_VOS_MODULE		= 0, /** version object store */
-	DAOS_MGMT_MODULE	= 1, /** storage management */
-	DAOS_POOL_MODULE	= 2, /** pool service */
-	DAOS_CONT_MODULE	= 3, /** container service */
-	DAOS_OBJ_MODULE		= 4, /** object service */
-	DAOS_REBUILD_MODULE	= 5, /** rebuild **/
-	DAOS_RSVC_MODULE	= 6, /** replicated service server */
-	DAOS_RDB_MODULE		= 7, /** rdb */
-	DAOS_RDBT_MODULE	= 8, /** rdb test */
-	DAOS_SEC_MODULE		= 9, /** security framework */
-	DAOS_DTX_MODULE		= 10, /** DTX */
-	DAOS_PIPELINE_MODULE	= 11,
-	DAOS_NR_MODULE		= 12, /** number of defined modules */
-	DAOS_MAX_MODULE		= 64  /** Size of uint64_t see dmg profile */
+#define X(a, b) a = b,
+	DAOS_MODULE_LIST
+#undef X
```

Comment:
why do we need to add explicit values for enums ? all enums start from 0

Reply:
> why do we need to add explicit values for enums ? all enums start from 0

in this case i am guessing cause some values are disjoint, like MAX being 64. Also with macros now, you need to either have all values being explicit or none in a macro.

You would need something like this, which I don't think is possible with conventional macros:
#define X(a, b) a = b (if b is set, else) a

Reply:
Sure I would certainly not try to complicate the macro. But in this case I honestly think that's a mistake and you should not bother having it in the macro either, DAOS_MAX_MODULE should not be 64, it should be the last enum value + 1. 
 
> some values

There is no other disjoint value :)

Reply:
If the last one is special then yes, there are better ways to handle it.

Reply:
As discussed in the last cart meeting, ideally this list should also include cart (as we occupy the same module space), and in carts case we reserve 0xff and 0xfe making the list disjoint. I want to unify those lists in the subsequent patch, which also aims at removing headers and storing resolved rpc strings instead in the opc (earlier suggestion by @ashleypittman and @liw )